### PR TITLE
feat: Min and max numerical aggregates

### LINKED ISSUE #2978: Support for `_min` and `_max` aggregates
We should add an aggregate for min and max values from an array of numerical values.

### PR DESCRIPTION
**Pull request by nasdf**
## Relevant issue(s)

Resolves #2978

## Description

This PR adds support for two new query aggregates `min` and `max`.

## Tasks

- [x] I made sure the code is well commented, particularly hard-to-understand areas.
- [x] I made sure the repository-held documentation is changed accordingly.
- [x] I made sure the pull request title adheres to the conventional commit style (the subset used in the project can be found in [tools/configs/chglog/config.yml](tools/configs/chglog/config.yml)).
- [x] I made sure to discuss its limitations such as threats to validity, vulnerability to mistake and misuse, robustness to invalidation of assumptions, resource requirements, ...

## How has this been tested?

Added integration tests

Specify the platform(s) on which this was tested:
- MacOS

### CODE REVIEW

File: tests/integration/query/simple/with_group_average_max_test.go
Comment (fredcarle) [inline tests/integration/query/simple/with_group_average_max_test.go:115]: typo: `_max`

**Review by AndrewSisley — CHANGES_REQUESTED** [review]
I want too review this but will wait until the tests are passing, please re-request my review when they are and I'll get right on this :)

File: tests/integration/query/simple/with_group_average_max_test.go
Comment (nasdf) [reply tests/integration/query/simple/with_group_average_max_test.go]: fixed

**Comment by nasdf** [comment]
> I want too review this but will wait until the tests are passing, please re-request my review when they are and I'll get right on this :)

Should be all fixed now. There was an issue with casting max value floats to ints which I think may still exist in the sum aggregate where I copied the code from.

**Review by AndrewSisley — CHANGES_REQUESTED** [review]
Looks good to me, just some minor stuff to sort out first IMO, and the below question.

question: It does not lok like you have added max/min to non-numeric types such as `DateTime`, `String`, `Boolean`, etc, and the gql related code does not look like it will facilitate the introduction of that support - was this an oversight or do you have a plan on how to introduce it later?  I am concerned that doing so will be a breaking change to the GQL syntax due to the return type.

**Review by fredcarle — APPROVED** [review]
Nice work! I like what you did with the aggregate generics and thanks for adding all those tests. Just need to resole Andy's comments before merge.

File: internal/planner/max.go
Comment (shahzadlone) [inline internal/planner/max.go:71]: suggestion: Would really like this to be hit with the tests

File: internal/planner/max.go
Comment (shahzadlone) [inline internal/planner/max.go:95]: todo: These two lines are not hit in any tests, if easy to do so, pelase add

File: internal/planner/max.go
Comment (shahzadlone) [inline internal/planner/max.go:103]: suggestion: same comment as above.

File: internal/planner/min.go
Comment (shahzadlone) [inline internal/planner/min.go:95]: suggestion: Similar comment as made on `max` explain.

File: internal/planner/min.go
Comment (shahzadlone) [inline internal/planner/min.go:103]: suggestion: similar comment as `max` explain

File: tests/integration/explain.go
Comment (shahzadlone) [inline tests/integration/explain.go:45]: praise: :)

File: tests/integration/explain/debug/top_with_max_test.go
Comment (shahzadlone) [inline tests/integration/explain/debug/top_with_max_test.go:34]: question: This is not under the `selectTopNode`? Is a sibling to `selectTopNode` might be worth documenting that at planner level if not already (maybe I missed it).

File: tests/integration/explain/debug/with_max_join_test.go
Comment (shahzadlone) [inline tests/integration/explain/debug/with_max_join_test.go:31]: question: here `maxNode` is under `selectTopNode` in the `debug` explain but not in the example above ? Why is that?

File: internal/planner/aggregate.go
Comment (nasdf) [reply internal/planner/aggregate.go]: done

File: internal/planner/max.go
Comment (nasdf) [reply internal/planner/max.go:54]: I'm not seeing which part is incorrect.

File: internal/planner/max.go
Comment (nasdf) [reply internal/planner/max.go:73]: added a test that should cover it

File: internal/planner/max.go
Comment (nasdf) [reply internal/planner/max.go]: refactored to return nil on empty values

Comment (nasdf) [reply internal/planner/max.go]: refactored, this is no longer applicable

File: internal/planner/max.go
Comment (nasdf) [reply internal/planner/max.go:33]: done

File: internal/planner/max.go
Comment (nasdf) [reply internal/planner/max.go:73]: should be covered now. I noticed that every other aggregate is also missing this coverage

File: tests/integration/explain/debug/with_max_join_test.go
Comment (nasdf) [reply tests/integration/explain/debug/with_max_join_test.go:31]: They are testing two different cases. This one is testing an aggregate from within a select and the other is testing an aggregate from the top level.

File: internal/planner/max.go
Comment (AndrewSisley) [inline internal/planner/max.go:147]: todo: This is incorrect for large integers, you cannot convert a very big int to a float and then back again and expect it to have the same value.

suggestion: Add failing tests before fixing this (if you can, you might struggle to set an Int to Int64MaxVal atm via our public clients, as GQL is technically only 32bit Ints)

File: tests/integration/explain/debug/top_with_max_test.go
Comment (nasdf) [reply tests/integration/explain/debug/top_with_max_test.go:34]: In this test case it's used from the top level.

File: internal/planner/max.go
Comment (AndrewSisley) [reply internal/planner/max.go:147]: e.g. the below will result in a negative number:
```
a := math.MaxInt64 - 1
		println(fmt.Sprint(a))
		b := float64(a)
		c := int64(b)
		panic(fmt.Sprint(c))
```

Comment (nasdf) [reply internal/planner/max.go]: refactored to use `big.Float`

File: internal/planner/max.go
Comment (nasdf) [reply internal/planner/max.go:97]: should be covered now

File: internal/planner/max.go
Comment (nasdf) [reply internal/planner/max.go:105]: should also be covered now

File: internal/planner/min.go
Comment (nasdf) [reply internal/planner/min.go:97]: should also be covered now

File: internal/planner/min.go
Comment (nasdf) [reply internal/planner/min.go:105]: should also be covered now

**Comment by nasdf** [comment]
> Which aggregate types? The new types added in this PR? Or some existing stuff to?

I haven't thought through all of it yet, but because the aggregates are expected to return a float and can then be used as input into other aggregates, there will be issues with an aggregate that can return a union type of String, DateTime, etc.
 
> I am concerned that you will just end up re-writing a lot the new stuff added in this PR to handle (IMO) a pretty core sub-feature (DateTime).

The plan node code is very generic. I'm not too worried about this.

**Review by shahzadlone — APPROVED** [review]
LGTM thanks for doing the suggestions

File: internal/planner/max.go
Comment (AndrewSisley) [reply internal/planner/max.go]: did you try to test this?

**Comment by AndrewSisley** [comment]
> there will be issues with an aggregate that can return a union type of String, DateTime, etc.

Yes, that is why I am concerned you are painting the syntax into a corner and we will need to make an avoidable breaking change in the (possibly near) future in order to correct this problem.  The breaking change will require us to change our tests, and users to change their code.

I do not think this is a good thing to do, but will not block the PR over it.  Please resolve the cast problem/test request before merge though.

File: internal/planner/max.go
Comment (nasdf) [reply internal/planner/max.go]: I've added tests for int max value

File: internal/planner/max.go
Comment (nasdf) [reply internal/planner/max.go]: Okay I see what you mean. I've refactored the logic so it actually returns the correct type now. 

Correct me if I'm wrong but I don't think we have any operations that would allow us to test mixed type aggregates.

Comment (AndrewSisley) [reply internal/planner/max.go]: > Correct me if I'm wrong but I don't think we have any operations that would allow us to test mixed type aggregates.

No you are correct, I'm only commenting on the comment, the code/tests are all good IMO :)

> I've refactored the logic so it actually returns the correct type now.

Nice, thanks!

**Comment by nasdf** [comment]
> > there will be issues with an aggregate that can return a union type of String, DateTime, etc.
> 
> Yes, that is why I am concerned you are painting the syntax into a corner and we will need to make an avoidable breaking change in the (possibly near) future in order to correct this problem. The breaking change will require us to change our tests, and users to change their code.
> 
> I do not think this is a good thing to do, but will not block the PR over it. Please resolve the cast problem/test request before merge though.

The GQL type system won't allow for an aggregate that returns a union to be used as input in another aggregate.

I think the best approach will be to add new aggregates: `_minDateTime`, `_maxString`, etc.

This shouldn't result in any breaking changes to the existing aggregates.

**Comment by AndrewSisley** [comment]
> I think the best approach will be to add new aggregates: _minDateTime, _maxString, etc.
>
> This shouldn't result in any breaking changes to the existing aggregates.

Only if we dont consider it weird that we have `_min` and not `_minFloat`, if that is the long term syntax we want I would prefer that all the types be consistent.

**Review by AndrewSisley — APPROVED** [review]
Talked over the long term plan in discord, is fully compatible with this PR - thanks for all of this Keenan, and for the discussion :)

**Comment by islamaliev** [comment]
bug bash result: all looks good.

**Pull request merged**
